File: finalists/Xavier/challenge_10/10.cpp
#include <iostream>
#include <stack>
using namespace std;

int number(string& s) {
  int sum = 0;
  for (int i = 0; i < (signed)s.size(); ++i) {
    sum *= 10;
    sum += s[i] - '0';
  } 
  return sum;
}

int main() {
  stack<int> st;
  string s; 
  while (cin >> s) {
      if (s == ".") cout << st.top() << endl;
      else {
	if (s == "mirror") {
	  int t = st.top(); st.pop();
	  st.push(-t);
	} else if (s == "breadandfish") {
	  int t = st.top();
	  st.push(t);
	} else if (s == "fire") {
	  st.pop();
	} else if (s == "$") {
	  int a = st.top(); st.pop(); 
	  int b = st.top(); st.pop();
	  st.push(b - a);
	} else if (s == "&") {
	  int a = st.top(); st.pop(); 
	  int b = st.top(); st.pop();
	  st.push(b/a);	  
	} else if (s == "@") {
	  int a = st.top(); st.pop(); 
	  int b = st.top(); st.pop();
	  st.push(b + a);
	} else if (s == "dance") {
	  int a = st.top(); st.pop(); 
	  int b = st.top(); st.pop();
	  st.push(a);
	  st.push(b);
	} else if (s == "#") {
	  int a = st.top(); st.pop(); 
	  int b = st.top(); st.pop();
	  st.push(b * a);
	} else if (s == "conquer") {
	  int a = st.top(); st.pop(); 
	  int b = st.top(); st.pop();
	  st.push(b % a);
	} else st.push(number(s));
	
      }
  }
}

File: finalists/Joan/challenge_10/10.py
#!/usr/bin/env python

import sys


#RPN
def operate(op, dig):
    

    #Print the solution
    if op == ".":
        print dig.pop()
        
    # -top
    elif op == "mirror":
        
        k = dig.pop()

        dig.append(-k)

    # replicate the top
    elif op == "breadandfish":
        k = dig.pop()
        dig.append(k)
        dig.append(k)
    # Remove the top
    elif op == "fire":
        a = dig.pop()
        #dig.append(a)
    # Multiplication
    elif op == "#":
        a = dig.pop()
        b = dig.pop()
        dig.append(a*b)

    # substraction
    elif op == "$":
        a = dig.pop()
        b = dig.pop()

        dig.append(b-a)
    # reverse
    elif op == "dance":
        dig.reverse()

    # divide and conquer ;)
    elif op == "conquer":
        a = dig.pop()
        b = dig.pop()
        dig.append(b%a)

    #Division
    elif op == "&":
        a = dig.pop()
        b = dig.pop()
        dig.append(b/a)

    #Addition
    elif op == "@":
        a = dig.pop()
        b = dig.pop()
        dig.append(a+b)

    
    else:
        #ERROR
        pass
        #print "ERROR"

if __name__ == '__main__':
    for s in sys.stdin:

   
            fields = s.split()

            dig = []

            for t in fields:
                if t.isdigit():
                    z = int(t)
                    dig.append(z)
                else:
                    operate(t,dig)
                    #k = []
    
